src/App/Fossa/Configuration.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module App.Fossa.Configuration
+( mergeFileCmdMetadata
+, readConfigFile
+, ConfigFile (..)
+, ConfigProject (..)
+, ConfigRevision (..)
+)
+where
+
+import Data.Aeson ( FromJSON(parseJSON), (.:), (.:?), withObject )
+import Data.Text (Text)
+import App.Types
+import Path
+import qualified Control.Carrier.Diagnostics as Diag
+import Effect.ReadFS
+import Options.Applicative
+
+data ConfigFile = ConfigFile
```

Comment:
I agree with you and I suspect Connor does as well since we were discussing some issues with how we manage configuration right now. This feels like a much bigger refactor than we should do in this PR to me since I think this would involve changing how we handle `CmdOptions` and configuration to all the entry points. That said, what parts of the `ConfigFile` type I created do you think would differ from `SpectrometerConfig`? I think it would be possible to make these identical so that we don't need to do this conversion.

Reply:
`SpectrometerConfig` should define a **valid configuration**, rather than a valid **partial** configuration, which the file format defines.  In `SpectrometerConfig`, values with hard-coded defaults would be actual values, not `Maybe`s.  We could also parse text values into their smarter data types, like `URI`s.

For example, `server` and `apiKey` should be `URI` and `ApiKey`, respectively.  `project.id` would be `Text`, instead of `Maybe Text`. `revision.commit` would be `Text` instead of `Maybe Text`.  Some of this is possible in the current state, but not all of the maybes can be removed.

Reply:
Resolving, this is work left for a follow-up refactor PR.